nengo_loihi/splitter.py

```diff
@@ -1,543 +1,141 @@
-            warnings.warn("No precomputable objects. Setting precompute=True "
-                          "has no effect.")
-    else:
-        assert len(networks.host_pre.all_objects) == 0, (
-            "Object erroneously added to host_pre")
-
-    return networks
-
-
-def place_nodes(networks):
-    # Only ChipReceiveNodes can be run on chip
-    for node in networks.original.all_nodes:
-        if isinstance(node, ChipReceiveNode):
-            # Typically ChipReceiveNodes are created by the splitter, but
-            # it's conceivable that advanced users might make them manually
-            networks.move(node, "chip")
+        # Step 5. Split precomputable parts of host
+        # This is a subset of host, marking which are precomputable
+        if precompute:
+            self._host_precomputable_objects = self._preclosure()
         else:
-            networks.move(node, "host")
-
-
-def place_ensembles(networks):
-    config = networks.original.config
-
-    for ens in networks.original.all_ensembles:
-        # User-specified config takes precedence
-        if config[ens].on_chip is not None:
-            networks.move(ens, "chip" if config[ens].on_chip else "host")
-        # Direct mode ensembles must be off-chip
-        elif isinstance(ens.neuron_type, Direct):
-            networks.move(ens, "host")
-
-    for conn in networks.original.all_connections:
-        # `post` of learning rules must be off chip
-        if (conn.learning_rule_type is not None
-                and isinstance(base_obj(conn.post_obj), Ensemble)
-                and conn.post_obj not in networks):
-            networks.move(conn.post_obj, "host")
-        # `error` of learning rules must be off chip
-        elif (isinstance(conn.post_obj, LearningRule)
-              and isinstance(base_obj(conn.pre_obj), Ensemble)
-              and conn.pre_obj not in networks):
-            networks.move(conn.pre_obj, "host")
-
-    # All other ensembles are placed on chip
-    for ens in networks.original.all_ensembles:
-        if ens not in networks:
-            networks.move(ens, "chip")
-
-
-def merge_passthrough_nodes(networks):
-    offchip = set()
-    for obj, target in networks.moves.items():
-        if isinstance(obj, Node) and obj.output is None:
-            # this is a passthrough Node so don't force it to be offchip
-            continue
-        elif target == 'host':
-            offchip.add(obj)
-
-    remove_nodes, remove_conns, add_conns = convert_passthroughs(
-        networks.original, offchip)
-    for n in remove_nodes:
-        networks.remove(n)
-    for c in remove_conns:
-        networks.remove(c)
+            self._host_precomputable_objects = set()
 
-    conns = networks.original.all_connections
-    for c in remove_conns:
-        conns.remove(c)
-    conns.extend(add_conns)
+    def _preclosure(self):  # noqa: C901
+        # performs a transitive closure on all host objects that
+        # send data to the chip
```

Comment:
What does this mean? Also, I'd make this a brief docstring.

Reply:
https://en.wikipedia.org/wiki/Transitive_closure#In_graph_theory

Will add docstring.